subsys/settings/settings_nvs.c

```diff
@@ -0,0 +1,229 @@
+	u16_t name_id = NVS_NAMECNT_ID;
+
+	/* In nvs the settings data is stored as:
+	 *	a. name at id starting from NVS_NAMECNT_ID (0 not included)
+	 *	b. value at id of name + NVS_NAME_ID_OFFSET
+	 * Deleted records will not be found, only the last record will be
+	 * read.
+	 */
+
+	name_id = cf->last_name_id + 1;
+
+	while (1) {
+
+		name_id--;
+		if (name_id == NVS_NAMECNT_ID) {
+			break;
+		}
+		rc1 = nvs_read(&cf->cf_nvs, name_id, &name, sizeof(name));
+		rc2 = nvs_read(&cf->cf_nvs, name_id + NVS_NAME_ID_OFFSET,
+			       &buf, sizeof(buf));
+
+		if ((rc1 <= 0) && (rc2 <= 0)) {
+			continue;
+		}
+
+		if ((rc1 <= 0) || (rc2 <= 0)) {
+			/* NVS is dirty, clean it */
+			if (name_id == cf->last_name_id) {
+				cf->last_name_id--;
+				nvs_write(&cf->cf_nvs, NVS_NAMECNT_ID,
+					  &cf->last_name_id, sizeof(u16_t));
+			}
+			nvs_delete(&cf->cf_nvs, name_id);
+			nvs_delete(&cf->cf_nvs, name_id + NVS_NAME_ID_OFFSET);
+			continue;
+		}
+
+		/* Found a name, this might not include a trailing \0 */
+		name[rc1] = '\0';
+		ch = settings_parse_and_lookup(name, &name_argc, name_argv);
+		if (!ch) {
+			continue;
+		}
+
+		/* Found a settings_handler for the name */
+		read_fn_arg.fs = &cf->cf_nvs;
+		read_fn_arg.id = name_id + NVS_NAME_ID_OFFSET;
+		ch->h_set(name_argc - 1, &name_argv[1], rc2,
+			  settings_nvs_read_fn, (void *) &read_fn_arg);
+	}
+	return 0;
+}
+
+/* ::csi_save implementation */
+static int settings_nvs_save(struct settings_store *cs, const char *name,
+			     const char *value, size_t val_len)
+{
+	struct settings_nvs *cf = (struct settings_nvs *)cs;
+	char rdname[SETTINGS_MAX_NAME_LEN + SETTINGS_EXTRA_LEN + 1];
+	u16_t name_id, write_name_id;
+	bool delete, write_name;
+	int rc = 0;
+
+	if (!name) {
+		return -EINVAL;
+	}
+
+	write_name = true;
+	/* Find out if we are doing a delete */
+	delete = false;
+	if ((value == NULL) || (val_len == 0)) {
+		delete = true;
+	}
+
+	name_id = cf->last_name_id + 1;
+	write_name_id = cf->last_name_id + 1;
+
+	while (1) {
+
+		name_id--;
+		if (name_id == NVS_NAMECNT_ID) {
+			break;
+		}
+
+		rc = nvs_read(&cf->cf_nvs, name_id, &rdname, sizeof(rdname));
+
+		if (rc < 0) {
+			/* Error or entry not found */
+			if (rc == -ENOENT) {
+				write_name_id = name_id;
+			}
+			continue;
+		}
+
+		rdname[rc] = '\0';
+
+		if (delete) {
+			if (strncmp(name, rdname, strlen(name))) {
+				continue;
+			}
+		} else {
+			if (strcmp(name, rdname)) {
+				continue;
+			}
+		}
+
+
+		if ((delete) && (name_id == cf->last_name_id)) {
+			cf->last_name_id--;
+			rc = nvs_write(&cf->cf_nvs, NVS_NAMECNT_ID,
+				       &cf->last_name_id, sizeof(u16_t));
+		}
+
+		if (delete) {
+			rc = nvs_delete(&cf->cf_nvs, name_id);
+			rc = nvs_delete(&cf->cf_nvs, name_id +
+					NVS_NAME_ID_OFFSET);
+			continue;
+		}
+		write_name_id = name_id;
+		write_name = false;
+		break;
+	}
+
+	if (delete) {
+		return rc;
+	}
+
+	/* write the value */
+	rc = nvs_write(&cf->cf_nvs, write_name_id + NVS_NAME_ID_OFFSET,
```

Comment:
I think it is possible to overwrite the last key if backend is at ID end.

Reply:
Yes this is true, but it does this when you have reached 16383 entries.